xtc/src/common_types.rs

```diff
@@ -0,0 +1,142 @@
+use derive_builder::*;
+use derive_new::*;
+use ic_kit::{
+    candid::{CandidType, Int, Nat},
+    Principal,
+};
+use std::convert::{TryFrom, TryInto};
+use xtc_history_common::types::*;
+
+type Time = Int;
+
+#[derive(CandidType, Clone)]
+pub enum Operation {
+    approve,
+    mint,
+    transfer,
+    transferFrom,
+    burn,
+    canisterCalled,
+    canisterCreated,
+}
+
+#[derive(CandidType)]
+pub struct Metadata<'a> {
+    pub decimals: u8,
+    pub fee: Nat,
+    pub logo: &'a str,
+    pub name: &'a str,
+    pub owner: Principal,
+    pub symbol: &'a str,
+    pub totalSupply: Nat,
+}
+
+#[derive(CandidType, Debug, Eq, PartialEq)]
+pub enum TxError {
+    InsufficientAllowance,
+    InsufficientBalance,
+}
+
+pub type TxReceipt = Result<Nat, TxError>;
+
+#[derive(CandidType, Clone, new)]
+pub struct TxRecord {
+    pub caller: Option<Principal>,
+    pub from: Principal,
+    pub to: Principal,
+    pub amount: Nat,
+    pub fee: Nat,
+    pub op: Operation,
+    pub timestamp: Time,
+    pub index: Nat,
+}
+
+impl TxRecord {
+    pub fn setIndex(mut self, index: Nat) -> Self {
+        self.index = index;
+        self
+    }
+}
+
+impl TryFrom<Transaction> for TxRecord {
```

Comment:
As I can see all of the cases are handled in the match expression, so why are you using `TryFrom` instead of `From`?

Reply:
that was not always the case, in the original implementation I deliberately left out some cases and wanted to treat them as errors, I would leave the TryFrom for greater flexibility in the future